### PR TITLE
k3s: package k3s-killall script

### PR DESCRIPTION
**Pull request by rorosen**
Provide the k3s-killall.sh script for orderly shutdown of k3s.

## Description of changes

This is an attempt to package [k3s-killall.sh](https://docs.k3s.io/upgrades/killall?_highlight=killall#killall-script), which is kinda weird as the killall script is generated by (and contained in) the k3s installer script. I copy the installer script and modify it so I can call only the `create_killall` function which generates the killall script. I'm not completely convinced by this approach tbh, but given the circumstances it's the "best" way I can think about. However, afaict the packaged script behaves the same way as it would in a regular k3s installation.

Additionally, I included the script in the single node test to verify that it's working. 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@Mic92 @euank @yajo @superherointj 

## Related Issue

- #98090 

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Mic92** [comment]
This could than replace this whole section in the wiki right? https://wiki.nixos.org/wiki/K3s#Cluster_Reset

**Comment by rorosen** [comment]
> This could than replace this whole section in the wiki right? https://wiki.nixos.org/wiki/K3s#Cluster_Reset

Not exactly, the killall script cleans up containers with associated data, networking components, iptables chains etc. The cluster data on server nodes (e.g. cluster datastore and token) will not be deleted.

I think the equivalent of what's described in the wiki would be [k3s-uninstall.sh](https://docs.k3s.io/installation/uninstall?_highlight=uninstall) but that's harder to package than the killall script.

**Comment by rorosen** [comment]
Just noticed that there are leftover containerd units by running `systemctl list-units|grep containerd` after running the killall script. This is different to a regular installation on ubuntu. I'll look into this and convert this PR to draft in the meantime.

**Comment by rorosen** [comment]
There are no more leftover units after killall and the test checks this now as well.

**Comment by thomasjm** [comment]
Thanks for this! Could we make it so that this script runs when you change `services.k3s.enable` from `true` to `false`?

IMO that would be the most Nix-y way for this to work. Hopefully there's an appropriate systemd hook that can run when the k3s service is removed.

**Comment by rorosen** [comment]
The idea is to just package killall so that people have it available. I don't think this should be run automatically if k3s is stopped, as k3s intentionally [keeps containers running when the K3s service is stopped](https://docs.k3s.io/upgrades/killall). However, you could hook it yourself in your config.

```nix
systemd.services = {
  k3s.onSuccess = [ "killall-k3s.service" ];
  killall-k3s.serviceConfig = {
    Type = "oneshot";
    ExecStart = "${k3s}/bin/k3s-killall.sh";
  };
};
``` 

Adding only a `ExecStop` or `ExecStopPost` hook to the k3s service won't work as killall also stops the k3s service. In fact, it stops every service file in `/etc/systemd/system/` that starts with `k3s` so take care that the killall service doesn't start with k3s.

**Comment by thomasjm** [comment]
Hmm, the k3s docs say that containers continue running when the service is stopped "to allow high availability during upgrades."

For the NixOS scenario, is there some way we can keep the containers running while you're updating to a new version of the k3s service, but also kill everything when the system is transitioning to a `services.k3s.enable=false` state? 

I feel it's important for a declarative system like NixOS that if I enable k3s, and then go back to disabling it, my system shouldn't end up in a different state with lots of extra containerd processes running (and these might even persist after a reboot?).

> won't work as killall also stops the k3s service

FWIW when I was experimenting with purging a k3s cluster from my system, I found that I needed to tweak it. It seems likely that we'd want to check in a slightly modified version of k3s-killall.sh in any case.

**Comment by rorosen** [comment]
The scope of this PR is to just package the upstream killall script, only making the required changes so that it works on NixOS the same way it would work in a regular k3s installation.

I don't think it's a big problem to require a manual cleanup after disabling k3s, but maybe this should be discussed in a separate issue.

**Review by euank — APPROVED** [review]
This lgtm! I appreciate the testing for it too!

The `sed`ing of the install script is both clever, and seems like it could be fragile if upstream changes that at lot, but we can also just wait and see if it does end up being a maintenance burden

Thank you!

**Comment by euank** [comment]
I'll throw in my own comments about trying to hook this into systemd lifecycle etc.

I think in one sense, it's very nice to make this all fully declarative and make it so disabling the k3s service cleans things up fully.

On the other hand, I think the killall script, as it's offered by upstream, is mainly suitable for manual human usage. It doesn't handle alternate interface names (if you did configure that), it removes all cni network namespaces, even if they're from a different piece of software, it deletes kubernetes iptables rules that likewise could be from another k8s tool.

Basically, it's a pretty blunt tool, and so it feels slightly scary to automatically run it.

It doesn't make it any better, but there are other units that don't fully cleanup, like for example `virtualisation.docker.enable = false;` also leaves behind running containers, iptables rules, etc.

**Review by thomasjm — APPROVED** [review]
Thanks for doing this!

One optional piece of feedback: I agree with @euank that the `sed` stuff looks fragile. I think it would be cleaner to just inline the script with the necessary changes for NixOS, for two reasons: 1) it's not very long, and 2) as previously mentioned, it makes some sweeping changes to the system so it might be better for maintainers to have the contents front and center so they can see what they're dealing with.

**Comment by superherointj** [comment]
> I'm happy to file a follow-up about hooking this into the systemd lifecycle.

@thomasjm @euank A compromise could be making it optional in the NixOS module.

**Comment by yajo** [comment]
In case it helps, I think [there are some packages or options that mark a NixOS upgrade as "it needs to reboot"](https://nixos.org/manual/nixos/stable/#sec-upgrading-automatic).

If that makes sense and is possible, you can set k3s to trigger the "needs reboot" state when disabled. Rebooting will also kill all, and after the reboot no new systemd units should start, so IMHO that would fix the non-declarativity problem.

I think that a module could enable k3s-killall on system shutdown, similar to what is explained in https://github.com/NixOS/nixpkgs/issues/98090#issuecomment-1980273467. That module should be separate from the k3s module itself, because otherwise, disabling k3s would disable the killall module and it wouldn't kill after disabled.

**Comment by rorosen** [comment]
Thanks for the reviews!

I'm also not super happy with the `sed`ing but copying the script would make it very possible to miss upstream changes. The script hasn't been changed upstream that often recently, but once in a while there are smaller adjustments. I also thought about patching the script, but I'm afraid that this may requires several diffs for different versions.

I tried to make the `sed` stuff less fragile by checking whether the expected patterns are really found and erroring if not. Still not stable but I hope that major bugs will be detected together with the tests.

**Review by yajo — APPROVED** [review]

**Review by superherointj — APPROVED** [review]
LGTM.

@rorosen Thanks for the contribution.
